tensorflow/go/operation.go

```diff
@@ -123,6 +148,67 @@ func (p Output) c() C.TF_Output {
 
 func (p Output) canBeAnInput() {}
 
+// NumConsumers returns the number of inputs that consume this output.
+func (p Output) NumConsumers() int {
+	return int(C.TF_OperationOutputNumConsumers(p.c()))
+}
+
+// Consumers returns the inputs that consume this output.
+func (p Output) Consumers() []OpInput {
+	max := p.NumConsumers()
+	consumers := make([]C.TF_Input, max)
+	n := C.TF_OperationOutputConsumers(p.c(), (*C.TF_Input)(unsafe.Pointer(&consumers[0])), C.int(max))
+	consumers = consumers[:int(n)]
+
+	var inputs []OpInput
+	for _, consumer := range consumers {
+		inputs = append(inputs, OpInput{
+			Index: int(consumer.index),
+			Op: &Operation{
+				c: consumer.oper,
+				g: p.Op.g,
+			},
+		})
+	}
+
+	return inputs
+}
+
+// OpInput is one of the inputs to an operation.
+type OpInput struct {
```

Comment:
I feel the naming combinations are getting a bit confusing. We now have three types - `Output`, `Input`, `OpInput`. `Operation.Output` returns an `Output`, `Output.Consumers()` returns an `OpInput`, `OpInput.Input` returns an `Output` etc. :)

I wonder if we can make this a little less confusing with some naming changes.
For example, would it make sense to call this type `Consumer` instead of `OpInput`?

Then we'd have something like:

```go
// Consumer identifies a node in the graph that consumes an Output.
type Consumer struct {
  // Op is the Operation that consumes an Output of another Operation.
  Op *Operation

  // Index identifies which input of Op is connected to the Output of another Operation.
  Index int
}
```

And `Output.Consumers` would return a `[]Consumer`.
Thoughts?

Reply:
I thought about naming it Consumer, but I wanted to keep it close to the C api types. Consumer sort of still has the same problem. `Operation.Input(n)` would return a Consumer which seems equally confusing. The issue is that there's a bidirectional link between inputs and outputs which makes it hard to name them cleanly. Might be better to rename `OpInput.Input` to be `OpInput.Producer`.

It's also equally confusing that there's a separate go `Input` type which doesn't match the C definition of Input.

Reply:
The C api method is named TF_OperationOutputConsumers and returns a TF_Input. But TF_OperationInput takes in an Input and returns an output.

I renamed it to be Consumers() to go from Output to Input, and Producer() to go from Input to Output.

```go
var op *Operation
op.Output(0).Consumers()
op.Input(0).Producer()
```

Reply:
I can understand the confusion and motivation here.
Having the go `Input` not match the C definition of `TF_Input` can be confusing, but at least it is only confusing to the developer of the Go library, not the user. Ideally, there would be no confusion anywhere :), but if there has to be, I'd say that less confusion for the Go API user trumps consistency with the underlying C API.

Can we do away with `Operation.Input()` altogether?

Reply:
(Seems our comments were in flight at the same time).
Good idea on the `Producer` and `Consumers` naming, I like it.

I'm still not convinced about `Operation.Input` though. It seems just as reasonable, or perhaps more so, for `Operation.Input` to return an `Input` object (i.e., the (Operation, Index) pair of the node that produces the i-th input for this operation).

Reply:
Could you clarify that last part? `Operation.Input(n)` does return the Input object for the i-th input of that operation.

https://github.com/tensorflow/tensorflow/pull/19915/files#diff-09716844b0f2a06173f8eed54ab5b398R89

Reply:
I thought about this a little more. Are you saying Input(n) should return an Output node? I.e. does what `Input(0).Producer()` does? That's reasonable. There's not really anything in OpInput that doesn't exist in Output.

Reply:
Yup, that's what I meant (i.e., it should return an `Input` implemented by `Output`, not an `OpInput`). Isn't that simple ;-)

Reply:
Or for now, just do away with the function altogether.

Reply:
If we get rid of OpInput, would that mean `*Operation.Output(0).Consumers()` return `[]*Operation`?

Reply:
@asimshankar bump since those messages got sent really close to eachother and I'm not sure if a notification got sent. Frustratring that GitHub doesn't live update these threads

Reply:
@d4l3k : I wasn't suggesting we get rid of `OpInput`. Here's my thinking (I'm only going to refer to the functions added in this PR)

- `Operation.NumInputs` and `Operation.InputListSize` stay exactly as they are.
- `Operation.Input(int)` goes away
-  `OpInput` becomes `Consumer`
- `Output.Consumers()` returns `[]Consumer`
- `Consumer.DataType` can stay (exactly like you ahve now) if you wish, or go, your call
- `Consumer.Producer` returns an `Output` (exactly like you have now)

Sound reasonable?

(I'm a bit hesitant about `Operation.Input(int)` because I'm not sure what it should return. If it is needed, perhaps it should return an `Input` since the i-th input of an operation could either be an `Output` or an `OutputList`?)

Reply:
I updated it as described. I still think `Operation.Input(int)` is valuable to have so you can traverse the graph in either direction, but I removed it from this PR. 

`InputList` and `OutputList` don't really exist in the C api. I can't think of a single situation under the current C api where `Operation.InputListSize` or `Operation.OutputListSize` would be useful. Are InputList/OutputList actually used anywhere? I googled named outputs/OutputList and couldn't really find anything.

Reply:
I'm fine with removing `InputListSize`, let's do that.
`OutputListSize` is used in the generated wrappers, for example:

https://github.com/tensorflow/tensorflow/blob/5fa7b03a255d3c0d05aa48e7604a94185ef6b9e2/tensorflow/go/op/wrappers.go#L30